# Data Preprocessing

Suppose you are assigned to develop a machine learning model to predict whether an individual earns more than USD 50,000 or less in a year using the 1994 US Census Data. The datasets are sourced from the UCI Machine Learning Repository at http://archive.ics.uci.edu/ml/datasets/Census+Income.

The repository provides 5 datasets. However, each dataset is raw and does not come in the form of ABT (Analytic Base Table). The datasets are apparently not ready for predictive modeling.

The objective of this notebook is to guide you through the data preprocessing steps on the raw datasets in a sequence of exercises. The expected outcome is "clean" data that can be directly fed into any machine learning algorithm within the Scikit-Learn Python module. The clean data should look like the dataset used in this case study on our website.

# Exercises

**Exercise 0**

Read the training and test datasets directly from the data URL's. Also, since the datasets do not contain the feature names, explicitly specify them while loading in the datasets. Once you read in adultData and adultTest datasets, concatenate them into a single dataset called df.

In [1]:
import pandas as pd
import numpy as np

In [2]:
#column names details are given in website 
col_names = ['age','workclass','fnlwgt','education','education-num','marital-status','occupation','relationship','race','sex','capital-gain','capital-loss','hours-per-week','native_country','income']

In [3]:
train_df = pd.read_csv("http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data",sep = ',', names=col_names)
train_df.head(2)

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   

        marital-status        occupation    relationship    race    sex  \
0        Never-married      Adm-clerical   Not-in-family   White   Male   
1   Married-civ-spouse   Exec-managerial         Husband   White   Male   

   capital-gain  capital-loss  hours-per-week  native_country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K

In [4]:
test_df = pd.read_csv("http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.test",
                      sep = ',', names= col_names, skiprows=1)
test_df.head(3)

age   workclass  fnlwgt    education  education-num       marital-status  \
0   25     Private  226802         11th              7        Never-married   
1   38     Private   89814      HS-grad              9   Married-civ-spouse   
2   28   Local-gov  336951   Assoc-acdm             12   Married-civ-spouse   

           occupation relationship    race    sex  capital-gain  capital-loss  \
0   Machine-op-inspct    Own-child   Black   Male             0             0   
1     Farming-fishing      Husband   White   Male             0             0   
2     Protective-serv      Husband   White   Male             0             0   

   hours-per-week  native_country   income  
0              40   United-States   <=50K.  
1              50   United-States   <=50K.  
2              40   United-States    >50K.

In [5]:
df = pd.concat([train_df,test_df])
df.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native_country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

 


**--------------------------------------------------------------------------------**

**`Another way to do this:`**

url = (
    "http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data",    
    "http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.test",
)  # pay attention that this defines a tuple that cannot be modified


`adultData` = pd.read_csv(url[0], sep = ',', names = col_names, header = None)

`adultTest` = pd.read_csv(url[1] , sep = ',', names = col_names, skiprows = 1)

`df` = pd.concat([adultData,adultTest])

df.head()

**--------------------------------------------------------------------------------**

**Exercise 1**

Make sure the feature types match the descriptions outlined in the Data Description section. For example, confirm `age` is a numeric feature.

In [6]:
df.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native_country    object
income            object
dtype: object

**Exercise 2**

Calculate the number of missing values for each feature. Does the result surprise you?

In [7]:
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native_country    0
income            0
dtype: int64

**Exercise 3**

In Exercise 2, you should see zero missing value for each value. This indicates some features are coded with different labels such as "?" and "99999", instead of NaN. To provide a better overview, generate summary statistics of df. Hint: Use the describe() method with `include=np.number` and `include=np.object`. Make sure you have Python 3.6+ for this to work!

In [8]:
# Only looking in numeric columns
df.describe(include=np.number).round(2)

age      fnlwgt  education-num  capital-gain  capital-loss  \
count  48842.00    48842.00       48842.00      48842.00       48842.0   
mean      38.64   189664.13          10.08       1079.07          87.5   
std       13.71   105604.03           2.57       7452.02         403.0   
min       17.00    12285.00           1.00          0.00           0.0   
25%       28.00   117550.50           9.00          0.00           0.0   
50%       37.00   178144.50          10.00          0.00           0.0   
75%       48.00   237642.00          12.00          0.00           0.0   
max       90.00  1490400.00          16.00      99999.00        4356.0   

       hours-per-week  
count        48842.00  
mean            40.42  
std             12.39  
min              1.00  
25%             40.00  
50%             40.00  
75%             45.00  
max             99.00

In [9]:
# Only looking in string value columns
df.describe(include=np.object)

<ipython-input-9-984795f0d1fa>:2: DeprecationWarning: `np.object` is a deprecated alias for the builtin `object`. To silence this warning, use `object` by itself. Doing this will not modify any behavior and is safe. 
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  df.describe(include=np.object)


workclass education       marital-status       occupation relationship  \
count      48842     48842                48842            48842        48842   
unique         9        16                    7               15            6   
top      Private   HS-grad   Married-civ-spouse   Prof-specialty      Husband   
freq       33906     15784                22379             6172        19716   

          race    sex  native_country  income  
count    48842  48842           48842   48842  
unique       5      2              42       4  
top      White   Male   United-States   <=50K  
freq     41762  32650           43832   24720

**Exercise 4**

In Exercise 3, you can see the target feature `income` has four unique values. This contradicts the definition of `income` as it should have two only labels: "<=50K" and ">50K". In this exercise, return the unique values of `income`.

In [10]:
df['income'].unique()

array([' <=50K', ' >50K', ' <=50K.', ' >50K.'], dtype=object)

**Exercise 5**

In Exercise 4, you should see income consists of 4 unique values. The values are `[' <=50K', ' >50K', ' <=50K.', ' >50K.']`. The value contains excessive white space. In this exercise:

Remove the excessive white space of income in df.
Correct the lable of income in df. In particular, relabel `>50K. and <=50K`. to `>50K and <=50K` respectively by removing .

In [11]:
df['income'] = df['income'].str.strip()
df['income'] = df['income'].str.replace('.','')

<ipython-input-11-623335e32634>:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  df['income'] = df['income'].str.replace('.','')


In [12]:
df['income'].unique()

array(['<=50K', '>50K'], dtype=object)

**Exercise 6**

In Exercise 5, you can see that the raw (or the pre-cleaned) income column contained excessive white spaces. Can other categorical features ('workclass', 'education', 'marital-status', 'relationship', 'occupation', 'race', 'sex','native-country') have the same problem? Check which features have excessive white spaces. Remove the white spaces if necessary.

In [13]:
Categorical_col = ['workclass','education','marital-status','occupation','relationship','race','sex','native_country']
for col in Categorical_col:
    print('Unique values of "'+col+'" before removing extra space')
    print(df[col].unique(),'\n')
    
    df[col]=df[col].str.strip()
    
    print('\n','Unique values of "'+col+'" after removing extra space::',)
    print(df[col].unique(),'\n')

Unique values of "workclass" before removing extra space
[' State-gov' ' Self-emp-not-inc' ' Private' ' Federal-gov' ' Local-gov'
 ' ?' ' Self-emp-inc' ' Without-pay' ' Never-worked'] 


 Unique values of "workclass" after removing extra space::
['State-gov' 'Self-emp-not-inc' 'Private' 'Federal-gov' 'Local-gov' '?'
 'Self-emp-inc' 'Without-pay' 'Never-worked'] 

Unique values of "education" before removing extra space
[' Bachelors' ' HS-grad' ' 11th' ' Masters' ' 9th' ' Some-college'
 ' Assoc-acdm' ' Assoc-voc' ' 7th-8th' ' Doctorate' ' Prof-school'
 ' 5th-6th' ' 10th' ' 1st-4th' ' Preschool' ' 12th'] 


 Unique values of "education" after removing extra space::
['Bachelors' 'HS-grad' '11th' 'Masters' '9th' 'Some-college' 'Assoc-acdm'
 'Assoc-voc' '7th-8th' 'Doctorate' 'Prof-school' '5th-6th' '10th'
 '1st-4th' 'Preschool' '12th'] 

Unique values of "marital-status" before removing extra space
[' Never-married' ' Married-civ-spouse' ' Divorced'
 ' Married-spouse-absent' ' Separated' ' 

**Exercise 7**

The `workclass, occupation,` and `native-country` contain some missing values encoded as "?". Check the percentage of "?" in each of workclass, occupation, and native-country.

In [14]:
missingvalcol = ['workclass','occupation','native_country']
for col in missingvalcol:
    print('Column Name : '+col)
    print(df[col].value_counts(normalize=True).mul(100).round(2),'\n')

Column Name : workclass
Private             69.42
Self-emp-not-inc     7.91
Local-gov            6.42
?                    5.73
State-gov            4.06
Self-emp-inc         3.47
Federal-gov          2.93
Without-pay          0.04
Never-worked         0.02
Name: workclass, dtype: float64 

Column Name : occupation
Prof-specialty       12.64
Craft-repair         12.51
Exec-managerial      12.46
Adm-clerical         11.49
Sales                11.27
Other-service        10.08
Machine-op-inspct     6.19
?                     5.75
Transport-moving      4.82
Handlers-cleaners     4.24
Farming-fishing       3.05
Tech-support          2.96
Protective-serv       2.01
Priv-house-serv       0.50
Armed-Forces          0.03
Name: occupation, dtype: float64 

Column Name : native_country
United-States                 89.74
Mexico                         1.95
?                              1.75
Philippines                    0.60
Germany                        0.42
Puerto-Rico                    0.3

**Exercise 8**

In Exercise 7, you will notice missing values for both of the workclass and occupation features are about 5.7%. The native-country feature contains less than 2% of missing values. Note that the missing data (around more than 90%) are predominantly for the <=50K income whereas ~76% of observations pertain to the <=50K income at an aggregate level. Therefore, we shall not to impute these missing values but remove them instead as there would be minimal information loss.

In this exercise, remove the rows where workclass="?", occupation="?" and native-country="?". Check that the number of observations reduces from 48,842 to 45,222.

In [15]:
df = df[df['workclass'] != '?']
df = df[df['occupation'] != '?']
df = df[df['native_country'] != '?']

df.shape

(45222, 15)

In [16]:
df['race'].value_counts(normalize=True).mul(100).round(2)

White                 86.03
Black                  9.35
Asian-Pac-Islander     2.88
Amer-Indian-Eskimo     0.96
Other                  0.78
Name: race, dtype: float64

**Exercise 9**

In Exercise 7, notice that native-country is too granular and unbalanced. That is, close to 90% of native-country is "United-States" and the remaining 10% is made up of 40 different countries. The granularity (or the large cardinality) would yield a large number of columns when we encode native-country. You should also notice race exhibits the same problem where white accounts more than 75% of instances. In this exercise,

- For `native-country`, relabel all other countries as "Other" except "United-States".
- Likewise, relabel all other `race` as "Other" except "White".

In [17]:
df.loc[df['native_country'] != 'United-States', 'native_country'] = 'other'
df.loc[df['race'] != 'White', 'race'] = 'other'

In [18]:
print('Race',df['race'].value_counts(),'\n')
print('Native country:',df['native_country'].value_counts())

Race White    38903
other     6319
Name: race, dtype: int64 

Native country: United-States    41292
other             3930
Name: native_country, dtype: int64


**Exercise 10**

Recall that fnlwgt stands for "Final Weight" defined by the US Census. The weight is an "estimate of the number of units in the target population that the responding unit represents" This feature aims to allocate similar weights to people with similar demographic characteristics. In short, fnlwght has no predictive power.

In this exercise, remove fnlwgt from df.

In [19]:
df = df.drop(columns='fnlwgt')
df.head(0)

Empty DataFrame
Columns: [age, workclass, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native_country, income]
Index: []

**Exercise 11**

We suspect education and education-num might carry the same information. If they represent the same information, we should remove one of them (why?). To see this, run len(df['education'].unique()) and len(df['education-num'].unique()). Both should give you 16 unique values. To vindicate our suspection, make sure education and education-num indeed represent the same information. Then, drop education from df.

**Hint:** try pd.pivot_table.

In [20]:
print('Education : ',len(df['education'].unique()),'\n')
print('Education Num : ',len(df['education-num'].unique()))

Education :  16 

Education Num :  16


In [21]:
pd.pivot_table(df[['education-num', 'education']], values='education-num',  index='education')

education-num
education                  
10th                      6
11th                      7
12th                      8
1st-4th                   2
5th-6th                   3
7th-8th                   4
9th                       5
Assoc-acdm               12
Assoc-voc                11
Bachelors                13
Doctorate                16
HS-grad                   9
Masters                  14
Preschool                 1
Prof-school              15
Some-college             10

In [22]:
df = df.drop(columns='education')
df.head(0)

Empty DataFrame
Columns: [age, workclass, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native_country, income]
Index: []

#### **Remark**

In the previous exercises, we have performed heaps of data wrangling on target and categorical features. Let's focus on the continuous/numeric features: age, capital-gain, capital-loss and hours-per-week. Based on the summary statistics (in Exercise 2), age ranges 17 to 90 years old with a mean of 38.64. Therefore, we can conclude age has a reasonable range of values and hence requires no data wrangling. However, hours-per-week ranges from 1 to 99 hours with a mean of 40. We suspect 99 is used to label the "missing" instance of hours-per-week. On the other hand, it is still possible to work more than 90 hours per week. Therefore, we shall not preprocess it further. To check this, you can find the second largest value of hours-per-week is 98. Next, capital-gain ranges from 0 to 99,999 and we suspect that the missing observations are hence labeled as "99999". Is that true? We run the following code and find that a capital gain value of 99,999 always returns a higher income earner, we conjecture it might be a useful predictive value and hence we shall not remove the observations with this value.

df.loc[df['capital-gain'] == 99999.000000, 'income'].value_counts()

In [23]:
df.loc[df['capital-gain'] == 99999.000000, 'income'].value_counts()

>50K    229
Name: income, dtype: int64

**Exercise 12**

We suspect that `capital-loss =  − capital-gain` because an individual can either pay any capital gain or claim for capital loss to reduce capital gain in future (TaxBracket.org, 2017). Another possibility is to pay neither gain nor loss. Hence, it is more reasonable to record gain or loss as a single variable, rather than having them separate. Before defining such new variable, define a "mask_both" to verify that no observations were recorded as both positive capital-gain and capital-loss values. You should see there is zero count for "True" in the "mask_both".

In [24]:
mask_both = (df['capital-gain']>0) & (df['capital-loss']>0)
mask_both.value_counts()

False    45222
dtype: int64

**Exercise 13**

In light of Exercise 12, define a variable named capital which is given as `capital-gain - capital-loss`. Then remove `capital-gain` and `capital-loss` from df.

In [25]:
df['capital'] = df['capital-gain'] - df['capital-loss']
df = df.drop(columns=['capital-gain','capital-loss'])
df['capital'].describe()

count    45222.000000
mean      1012.834925
std       7530.315380
min      -4356.000000
25%          0.000000
50%          0.000000
75%          0.000000
max      99999.000000
Name: capital, dtype: float64

**Exercise 14**

As a good practice, the target feature should be the last column in the data frame. Move the target feature income to the end.

In [26]:
df.columns

Index(['age', 'workclass', 'education-num', 'marital-status', 'occupation',
       'relationship', 'race', 'sex', 'hours-per-week', 'native_country',
       'income', 'capital'],
      dtype='object')

In [27]:
cols = list(df.columns.values) #Make a list of all of the columns in the df
cols.pop(cols.index('income')) #Remove target feature from list
df = df[cols+['income']] #Create new dataframe with columns in the order you want

df.columns

Index(['age', 'workclass', 'education-num', 'marital-status', 'occupation',
       'relationship', 'race', 'sex', 'hours-per-week', 'native_country',
       'capital', 'income'],
      dtype='object')

**Exercise 15**

Print the shape of df and a list of the columns at this point. Also randomly sample 4 rows with a random state of 11.


In [28]:
df.shape

(45222, 12)

In [29]:
print('Shape : ',df.shape)
print('\n','Column List : ', df.columns)
df.sample(n=4, random_state=11)

Shape :  (45222, 12)

 Column List :  Index(['age', 'workclass', 'education-num', 'marital-status', 'occupation',
       'relationship', 'race', 'sex', 'hours-per-week', 'native_country',
       'capital', 'income'],
      dtype='object')


age     workclass  education-num      marital-status      occupation  \
10179   43  Self-emp-inc             16  Married-civ-spouse  Prof-specialty   
16095   45       Private             10            Divorced    Adm-clerical   
2006    57       Private             10       Never-married    Adm-clerical   
31231   48       Private             10  Married-civ-spouse           Sales   

         relationship   race     sex  hours-per-week native_country  capital  \
10179            Wife  White  Female              70  United-States        0   
16095   Not-in-family  White  Female              40  United-States        0   
2006   Other-relative  other  Female              40          other        0   
31231         Husband  other    Male              40  United-States        0   

      income  
10179   >50K  
16095  <=50K  
2006   <=50K  
31231   >50K

**Exercise 16**

Remove the income feature from the full dataset and call it target. Get a distribution of the values of the target feature (that is, value counts). Name the rest of the features Data, which are your descriptive features.



In [30]:
target = df['income']

Data = df.drop(columns='income')

print(target.value_counts())

<=50K    34014
>50K     11208
Name: income, dtype: int64


**Exercise 17**

Label-encode the target feature so that the positive class is ">50K" and it is encoded as "1". The negative class should be encoded as "0". Confirm correctness of your label-encoding by getting a value counts.

In [31]:
target = target.replace({'<=50K': 0, '>50K': 1})

target.value_counts()

0    34014
1    11208
Name: income, dtype: int64

**Exercise 18**

Get a list of all descriptive categorical features. Display value counts for each one of these features. Comment on which feature(s) appear to be ordinal.

In [32]:
categorical_cols = Data.columns[Data.dtypes == np.object].tolist()
categorical_cols

<ipython-input-32-0cad41afccff>:1: DeprecationWarning: `np.object` is a deprecated alias for the builtin `object`. To silence this warning, use `object` by itself. Doing this will not modify any behavior and is safe. 
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  categorical_cols = Data.columns[Data.dtypes == np.object].tolist()


['workclass',
 'marital-status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'native_country']

In [33]:
for categorical_col in categorical_cols:
    print(categorical_col + ':')
    print(Data[categorical_col].value_counts())
    print('\n')

workclass:
Private             33307
Self-emp-not-inc     3796
Local-gov            3100
State-gov            1946
Self-emp-inc         1646
Federal-gov          1406
Without-pay            21
Name: workclass, dtype: int64


marital-status:
Married-civ-spouse       21055
Never-married            14598
Divorced                  6297
Separated                 1411
Widowed                   1277
Married-spouse-absent      552
Married-AF-spouse           32
Name: marital-status, dtype: int64


occupation:
Craft-repair         6020
Prof-specialty       6008
Exec-managerial      5984
Adm-clerical         5540
Sales                5408
Other-service        4808
Machine-op-inspct    2970
Transport-moving     2316
Handlers-cleaners    2046
Farming-fishing      1480
Tech-support         1420
Protective-serv       976
Priv-house-serv       232
Armed-Forces           14
Name: occupation, dtype: int64


relationship:
Husband           18666
Not-in-family     11702
Own-child          6626
Unmarried 

**Exercise 19**

So all the categorical features appear to be nominal. Perform one-hot encoding for all the descriptive categorical features and call this encoded data frame as `Data_encoded`. If a categorical descriptive feature has only 2 levels, encode it with only one binary variable. For other categorical features (with more than 2 levels), use regular one-hot-encoding (where number of binary variables are equal to the number of distinct levels).

In [34]:
Data_encoded = Data.copy()  # retain original Data without encoding for further analysis

# if a categorical descriptive feature has only 2 levels,
# define only one binary variable
for col in categorical_cols:
    q = len(Data_encoded[col].unique())
    if (q==2):
        Data_encoded[col] = pd.get_dummies(Data_encoded[col], drop_first=True)
        
# for other categorical features (with > 2 levels), 
# perform regular one-hot-encoding using pd.get_dummies()
# if a feature is numeric, it will be untouched
Data_encoded = pd.get_dummies(Data_encoded)

**Exercise 20**

How many descriptive features are there now? Randomly sample 4 rows with a random state of 11.

In [35]:
print(f'Shape of the Data Enconded is {Data_encoded.shape}\n')

Data_encoded.sample(n=5,random_state=11)

Shape of the Data Enconded is (45222, 41)



age  education-num  race  sex  hours-per-week  native_country  capital  \
10179   43             16     0    0              70               0        0   
16095   45             10     0    0              40               0        0   
2006    57             10     1    0              40               1        0   
31231   48             10     1    1              40               0        0   
25952   48              9     0    1              50               0        0   

       workclass_Federal-gov  workclass_Local-gov  workclass_Private  ...  \
10179                      0                    0                  0  ...   
16095                      0                    0                  1  ...   
2006                       0                    0                  1  ...   
31231                      0                    0                  1  ...   
25952                      0                    0                  0  ...   

       occupation_Protective-serv  occupation_Sales  occupation_Tech-support  \
10179                           0                 0                        0   
16095                           0                 0                        0   
2006                            0                 0                        0   
31231                           0                 1                        0   
25952                           0                 1                        0   

       occupation_Transport-moving  relationship_Husband  \
10179                            0                     0   
16095                            0                     0   
2006                             0                     0   
31231                            0                     1   
25952                            0                     1   

       relationship_Not-in-family  relationship_Other-relative  \
10179                           0                            0   
16095                           1                            0   
2006                            0                            1   
31231                           0                            0   
25952                           0                            0   

       relationship_Own-child  relationship_Unmarried  relationship_Wife  
10179                       0                       0                  1  
16095                       0                       0                  0  
2006                        0                       0                  0  
31231                       0                       0                  0  
25952                       0                       0                  0  

[5 rows x 41 columns]

**Exercise 21**

After encoding all the categorical features, we end up with a data frame that is all numerical. Get a description of `Data_encoded` with include='all' option.

Next, perform a range normalization of the descriptive features using `MinMaxScaler()` method within `preprocessing` submodule of `Scikit-Learn`, and call it `Data_encoded_norm_numpy`. But make sure you keep `Data_encoded` around to keep track of column names.

In [36]:
Data_encoded.describe(include='all').round(3)

age  education-num       race        sex  hours-per-week  \
count  45222.000      45222.000  45222.000  45222.000       45222.000   
mean      38.548         10.118      0.140      0.675          40.938   
std       13.218          2.553      0.347      0.468          12.008   
min       17.000          1.000      0.000      0.000           1.000   
25%       28.000          9.000      0.000      0.000          40.000   
50%       37.000         10.000      0.000      1.000          40.000   
75%       47.000         13.000      0.000      1.000          45.000   
max       90.000         16.000      1.000      1.000          99.000   

       native_country    capital  workclass_Federal-gov  workclass_Local-gov  \
count       45222.000  45222.000              45222.000            45222.000   
mean            0.087   1012.835                  0.031                0.069   
std             0.282   7530.315                  0.174                0.253   
min             0.000  -4356.000                  0.000                0.000   
25%             0.000      0.000                  0.000                0.000   
50%             0.000      0.000                  0.000                0.000   
75%             0.000      0.000                  0.000                0.000   
max             1.000  99999.000                  1.000                1.000   

       workclass_Private  ...  occupation_Protective-serv  occupation_Sales  \
count          45222.000  ...                   45222.000         45222.000   
mean               0.737  ...                       0.022             0.120   
std                0.441  ...                       0.145             0.324   
min                0.000  ...                       0.000             0.000   
25%                0.000  ...                       0.000             0.000   
50%                1.000  ...                       0.000             0.000   
75%                1.000  ...                       0.000             0.000   
max                1.000  ...                       1.000             1.000   

       occupation_Tech-support  occupation_Transport-moving  \
count                45222.000                    45222.000   
mean                     0.031                        0.051   
std                      0.174                        0.220   
min                      0.000                        0.000   
25%                      0.000                        0.000   
50%                      0.000                        0.000   
75%                      0.000                        0.000   
max                      1.000                        1.000   

       relationship_Husband  relationship_Not-in-family  \
count             45222.000                   45222.000   
mean                  0.413                       0.259   
std                   0.492                       0.438   
min                   0.000                       0.000   
25%                   0.000                       0.000   
50%                   0.000                       0.000   
75%                   1.000                       1.000   
max                   1.000                       1.000   

       relationship_Other-relative  relationship_Own-child  \
count                     45222.00               45222.000   
mean                          0.03                   0.147   
std                           0.17                   0.354   
min                           0.00                   0.000   
25%                           0.00                   0.000   
50%                           0.00                   0.000   
75%                           0.00                   0.000   
max                           1.00                   1.000   

       relationship_Unmarried  relationship_Wife  
count               45222.000          45222.000  
mean                    0.106              0.046  
std                     0.308              0.210  
min                     0.000              0.000  
25%     

In [37]:
from sklearn import preprocessing

Data_scaler = preprocessing.MinMaxScaler()

Data_encoded_norm_numpy = Data_scaler.fit_transform(Data_encoded)

**Exercise 22**

Pay attention that the output of the scaler is a NumPy array, so all the column names are lost. That's why you kept a copy of `Data_encoded` before scaling so that you can recover the column names. Define a new Pandas data frame called `Data_encoded_norm_df` from `Data_encoded_norm_numpy` with the column names of `Data_encoded`.

Then have another look at the descriptive features after scaling by randomly sampling 4 rows with a random state of 11 from `Data_encoded_norm_df`.

Finally, get the shape and a description of `Data_encoded_norm_df` with include='all' option. Observe that everything is now between 0 and 1 and that binary features are still kept as binary after the min-max scaling.

In [38]:
Data_encoded_norm_df = pd.DataFrame(Data_encoded_norm_numpy, columns=Data_encoded.columns)

Data_encoded_norm_df.sample(n=5,random_state=11)

age  education-num  race  sex  hours-per-week  native_country  \
9406   0.356164       1.000000   0.0  0.0        0.704082             0.0   
45043  0.383562       0.600000   0.0  0.0        0.397959             0.0   
1848   0.547945       0.600000   1.0  0.0        0.397959             1.0   
28941  0.424658       0.600000   1.0  1.0        0.397959             0.0   
24047  0.424658       0.533333   0.0  1.0        0.500000             0.0   

        capital  workclass_Federal-gov  workclass_Local-gov  \
9406   0.041742                    0.0                  0.0   
45043  0.041742                    0.0                  0.0   
1848   0.041742                    0.0                  0.0   
28941  0.041742                    0.0                  0.0   
24047  0.041742                    0.0                  0.0   

       workclass_Private  ...  occupation_Protective-serv  occupation_Sales  \
9406                 0.0  ...                         0.0               0.0   
45043                1.0  ...                         0.0               0.0   
1848                 1.0  ...                         0.0               0.0   
28941                1.0  ...                         0.0               1.0   
24047                0.0  ...                         0.0               1.0   

       occupation_Tech-support  occupation_Transport-moving  \
9406                       0.0                          0.0   
45043                      0.0                          0.0   
1848                       0.0                          0.0   
28941                      0.0                          0.0   
24047                      0.0                          0.0   

       relationship_Husband  relationship_Not-in-family  \
9406                    0.0                         0.0   
45043                   0.0                         1.0   
1848                    0.0                         0.0   
28941                   1.0                         0.0   
24047                   1.0                         0.0   

       relationship_Other-relative  relationship_Own-child  \
9406                           0.0                     0.0   
45043                          0.0                     0.0   
1848                           1.0                     0.0   
28941                          0.0                     0.0   
24047                          0.0                     0.0   

       relationship_Unmarried  relationship_Wife  
9406                      0.0                1.0  
45043                     0.0                0.0  
1848                      0.0                0.0  
28941                     0.0                0.0  
24047                     0.0                0.0  

[5 rows x 41 columns]

In [39]:
print(f'Shape of Data_encoded_norm_df is {Data_encoded_norm_df.shape}\n')

Data_encoded_norm_df.describe(include='all').round(3)

Shape of Data_encoded_norm_df is (45222, 41)



age  education-num       race        sex  hours-per-week  \
count  45222.000      45222.000  45222.000  45222.000       45222.000   
mean       0.295          0.608      0.140      0.675           0.408   
std        0.181          0.170      0.347      0.468           0.123   
min        0.000          0.000      0.000      0.000           0.000   
25%        0.151          0.533      0.000      0.000           0.398   
50%        0.274          0.600      0.000      1.000           0.398   
75%        0.411          0.800      0.000      1.000           0.449   
max        1.000          1.000      1.000      1.000           1.000   

       native_country    capital  workclass_Federal-gov  workclass_Local-gov  \
count       45222.000  45222.000              45222.000            45222.000   
mean            0.087      0.051                  0.031                0.069   
std             0.282      0.072                  0.174                0.253   
min             0.000      0.000                  0.000                0.000   
25%             0.000      0.042                  0.000                0.000   
50%             0.000      0.042                  0.000                0.000   
75%             0.000      0.042                  0.000                0.000   
max             1.000      1.000                  1.000                1.000   

       workclass_Private  ...  occupation_Protective-serv  occupation_Sales  \
count          45222.000  ...                   45222.000         45222.000   
mean               0.737  ...                       0.022             0.120   
std                0.441  ...                       0.145             0.324   
min                0.000  ...                       0.000             0.000   
25%                0.000  ...                       0.000             0.000   
50%                1.000  ...                       0.000             0.000   
75%                1.000  ...                       0.000             0.000   
max                1.000  ...                       1.000             1.000   

       occupation_Tech-support  occupation_Transport-moving  \
count                45222.000                    45222.000   
mean                     0.031                        0.051   
std                      0.174                        0.220   
min                      0.000                        0.000   
25%                      0.000                        0.000   
50%                      0.000                        0.000   
75%                      0.000                        0.000   
max                      1.000                        1.000   

       relationship_Husband  relationship_Not-in-family  \
count             45222.000                   45222.000   
mean                  0.413                       0.259   
std                   0.492                       0.438   
min                   0.000                       0.000   
25%                   0.000                       0.000   
50%                   0.000                       0.000   
75%                   1.000                       1.000   
max                   1.000                       1.000   

       relationship_Other-relative  relationship_Own-child  \
count                     45222.00               45222.000   
mean                          0.03                   0.147   
std                           0.17                   0.354   
min                           0.00                   0.000   
25%                           0.00                   0.000   
50%                           0.00                   0.000   
75%                           0.00                   0.000   
max                           1.00                   1.000   

       relationship_Unmarried  relationship_Wife  
count               45222.000          45222.000  
mean                    0.106              0.046  
std                     0.308              0.210  
min                     0.000              0.000  
25%     

**Exercise 23**

Define a new data frame called `df_clean` which is the combination of the normalized and scaled descriptive features and the target feature with the target feature as the last column. Hint: Use assign(), but make sure you use the `.values` on the `target feature`.

Randomly sample 4 rows from `df_clean` with a random state of 11.

Write the clean dataset `df_clean` to a CSV file called `df_clean.csv`.

Finally, open this CSV file with your favorite CSV program (Excel should work fine) and have a look to make sure everything looks OK.

In [40]:
df_clean = Data_encoded_norm_df.assign(target = target.values)

df_clean.sample(n=4,random_state=11)

age  education-num  race  sex  hours-per-week  native_country  \
9406   0.356164            1.0   0.0  0.0        0.704082             0.0   
45043  0.383562            0.6   0.0  0.0        0.397959             0.0   
1848   0.547945            0.6   1.0  0.0        0.397959             1.0   
28941  0.424658            0.6   1.0  1.0        0.397959             0.0   

        capital  workclass_Federal-gov  workclass_Local-gov  \
9406   0.041742                    0.0                  0.0   
45043  0.041742                    0.0                  0.0   
1848   0.041742                    0.0                  0.0   
28941  0.041742                    0.0                  0.0   

       workclass_Private  ...  occupation_Sales  occupation_Tech-support  \
9406                 0.0  ...               0.0                      0.0   
45043                1.0  ...               0.0                      0.0   
1848                 1.0  ...               0.0                      0.0   
28941                1.0  ...               1.0                      0.0   

       occupation_Transport-moving  relationship_Husband  \
9406                           0.0                   0.0   
45043                          0.0                   0.0   
1848                           0.0                   0.0   
28941                          0.0                   1.0   

       relationship_Not-in-family  relationship_Other-relative  \
9406                          0.0                          0.0   
45043                         1.0                          0.0   
1848                          0.0                          1.0   
28941                         0.0                          0.0   

       relationship_Own-child  relationship_Unmarried  relationship_Wife  \
9406                      0.0                     0.0                1.0   
45043                     0.0                     0.0                0.0   
1848                      0.0                     0.0                0.0   
28941                     0.0                     0.0                0.0   

       target  
9406        1  
45043       0  
1848        0  
28941       1  

[4 rows x 42 columns]

In [41]:
# set index to False so that row IDs are not written
df_clean.to_csv('df_clean.csv',index=False)

**Exercise 24**

Scikit-Learn models require input data separately as the set of descriptive features and the target feature respectively. In addition, Scikit-Learn models require all data to be NumPy arrays (2-dimensional arrays for descriptive features and 1-dimensional arrays for the target feature). How would you go about specifying these two inputs with the objects you have defined so far?

In [42]:
# input descriptive features compatible with Scikit-learn models:

# Data_encoded_norm_numpy is already a NumPy array,
# it is also one-hot-encoded and scaled with no missing values,
# so it's good to go
Data_sk_learn = Data_encoded_norm_numpy

print(f'type of the target object is {type(target)}')

type of the target object is <class 'pandas.core.series.Series'>


In [43]:
# target object is a Pandas series, so you need to convert it 
# to a NumPy arrary via the .values method
target_sk_learn = target.values

print(f'type of the target object is {type(target_sk_learn)}')

type of the target object is <class 'numpy.ndarray'>


_Now everthing is done in preprocessing part, now these data is good to go for ML models._